### PR TITLE
xDS: Include orca named_metrics in LRS load reports

### PR DESCRIPTION
**Pull request by yousukseung**

### CODE REVIEW

File: include/grpcpp/ext/call_metric_recorder.h
Comment (yousukseung) [inline include/grpcpp/ext/call_metric_recorder.h:89]: > Let's just call this RecordNamedMetric().
Renamed.

File: src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_impl.cc
Comment (yousukseung) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_impl.cc:339]: > args.backend_metric_accessor will never be null. Also, it seems cleaner to not call GetBackendMetricData() multiple times.

Removed the null check and redundant calls.

File: src/core/ext/xds/xds_client_stats.h
Comment (yousukseung) [inline src/core/ext/xds/xds_client_stats.h:219]: > Suggest making this parameter a reference instead of a pointer.

I switched to a pointer to make it nullable so that I can avoid allocating an empty map when there's no load report, which I think is worth it here. From what I read GetBackendMetricData() wil return null if "endpoint-load-metrics-bin" trailing metadata is missing.

File: src/core/ext/xds/xds_client_stats.cc
Comment (yousukseung) [inline src/core/ext/xds/xds_client_stats.cc:134]: > No need for the = {}.
I don't see we initialize Snapshot fields, so believe total_* counters are undefined today. Instead I updated to explicitly initialize fields in the definition, I see no reason not to.

File: src/core/ext/xds/xds_client_stats.cc
Comment (yousukseung) [inline src/core/ext/xds/xds_client_stats.cc:145]: > These two lines should go in their own nested scope, so that we're not still holding the lock while we add the per-CPU snapshot to the aggregated snapshot on line 146.

Fixed.

File: src/core/lib/gprpp/per_cpu.h
Comment (yousukseung) [inline src/core/lib/gprpp/per_cpu.h:34]: > What's the purpose of this change? If there are actually more CPUs, don't we actually want that many more instances?
> If we do impose a limit on the total number of instances, what happens if this_cpu() gets called from a CPU that is greater than the limit? It seems like that would just cause a crash.

Thanks for catching this, fixed. Craig was concerned when there are too many cores. We actually did the same in Fathom in prod where we capped it at 16~32, so I picked 32 here, which I think will be fine.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h
Comment (yousukseung) [inline test/cpp/end2end/xds/xds_end2end_test_lib.h:786]: > I think this approach is inconsistent with the existing API pattern here...

Updated

File: test/cpp/end2end/xds/xds_cluster_end2end_test.cc
Comment (yousukseung) [inline test/cpp/end2end/xds/xds_cluster_end2end_test.cc:1437]: > Instead of having two maps here, I suggest a single map with both values:

Updated.

File: test/cpp/end2end/xds/xds_cluster_end2end_test.cc
Comment (yousukseung) [inline test/cpp/end2end/xds/xds_cluster_end2end_test.cc:1478]: Empty line removed.

File: test/cpp/end2end/xds/xds_cluster_end2end_test.cc
Comment (yousukseung) [inline test/cpp/end2end/xds/xds_cluster_end2end_test.cc:1612]: > Why not just replace the ::testing::_ with ::testing::IsEmpty(), so that you don't need a separate expectation immediately below?

Updated.

File: src/core/ext/xds/xds_client_stats.h
Comment (yousukseung) [reply src/core/ext/xds/xds_client_stats.h:219]: But it requires allocating an empty string unnecessarily unless we make this nullable in some way.
In XdsClusterImplLb::Picker::SubchannelCallTracker::Finish() where we call AddCallFinished():

```
auto* backend_metric_data = args.backend_metric_accessor->GetBackendMetricData();
const std::map<absl::string_view, double>* named_metrics = nullptr;
if (backend_metric_data != nullptr) {
  named_metrics = &backend_metric_data->named_metrics;
}
locality_stats_->AddCallFinished(named_metrics, !args.status.ok());
```

vs. Something like (pseudo-code)
```
auto* backend_metric_data = args.backend_metric_accessor->GetBackendMetricData();
std::map<absl::string_view, double> named_metrics
if (backend_metric_data == nullptr) {
  locality_stats_->AddCallFinished(std::map<..>(), !args.status.ok());
} else {
  locality_stats_->AddCallFinished(
    backend_metric_data->named_metrics, !args.status.ok());
}
```

File: src/core/ext/xds/xds_client_stats.h
Comment (yousukseung) [reply src/core/ext/xds/xds_client_stats.h]: Done.

File: src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc
Comment (yousukseung) [reply src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc:1]: Yes, will merge after #32657. For future PRs, what's the best practice to chain PRs in github, or do we discourage it?

File: test/cpp/end2end/xds/xds_end2end_test_lib.h
Comment (yousukseung) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h]: Removed.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h
Comment (yousukseung) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h]: Updated. We want to copy each time for sequential tests, so made a const ref of an optional.

File: test/cpp/end2end/xds/xds_cluster_end2end_test.cc
Comment (yousukseung) [reply test/cpp/end2end/xds/xds_cluster_end2end_test.cc]: Done.

File: test/cpp/end2end/xds/xds_cluster_end2end_test.cc
Comment (yousukseung) [reply test/cpp/end2end/xds/xds_cluster_end2end_test.cc]: Done.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h
Comment (markdroth) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h]: I don't understand the benefit of using a const reference here.

There are two possible cases for the caller:
1. They want to retain their own copy of the data they pass in.
2. They do not need to retain their own copy of the data they pass in.

In case 1, it does not matter if we use a const reference or not, because either way we will make our own copy.  If we use a const reference, we make the copy inside of the method where we assign to the data member; if we take the parameter by value, the copy is made when the method is invoked, and the method's body can use `std::move()` to move that copy to the data member.

In case 2, if we take the argument by value, then the caller can use `std::move()` when they invoke the method, and we can still use `std::move()` inside the method, so there is no copy at all.  However, if we use a const reference, then we always make a copy, even though we don't actually need one.

This isn't a big deal either way in this case, but it seems strictly better to pass it by value.  I'll leave it up to you.

File: src/core/ext/xds/xds_client_stats.h
Comment (yousukseung) [reply src/core/ext/xds/xds_client_stats.h:219]: I'll keep it then. Certainly no dynamic allocation, but I found this option looks slightly better: https://screenshot.googleplex.com/B5csTZpq6Eda2Gv

File: src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc
Comment (yousukseung) [reply src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc:1]: Ack.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h
Comment (yousukseung) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h]: Makes sense, updated.

File: test/cpp/end2end/xds/xds_cluster_end2end_test.cc
Comment (yousukseung) [reply test/cpp/end2end/xds/xds_cluster_end2end_test.cc:1378]: Done.

File: src/core/ext/xds/xds_client_stats.h
Comment (markdroth) [reply src/core/ext/xds/xds_client_stats.h:219]: Note that the first version in that screenshot is not the same as the version that I proposed in my last comment.  The version you're using imposes the overhead of a `std::move()`, which is not present in the version I proposed, so I think it should have less overhead.

In any case, though, as I said, this isn't worth holding up this PR over.

File: src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc
Comment (yousukseung) [reply src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc]: Something went wrong during `merge master` which gave me a "local copy is outdated" error, which I had to force-push. I think this change was related. Fixed and ack on force-push, I think in this case I can `merge origin` to avoid force-pushing later.

File: src/core/ext/xds/xds_client_stats.h
Comment (yousukseung) [reply src/core/ext/xds/xds_client_stats.h:219]: Right that we should only compare the empty case here, and it comes down to https://screenshot.googleplex.com/329NtyXoDS7yPZX. That said, I overestimated the cost of allocating an empty map in the stack which should be very cheap, however like you said I think we can keep it, which won't hurt. I will resolve thisl.

**Comment by yousukseung** [comment]
The EPS PR #32657 has been rolled back, will hold until we roll it forward.

**Comment by yousukseung** [comment]
> The EPS PR #32657 has been rolled back, will hold until we roll it forward.

@markdroth PR is now fully ready (pending tests).

**Review by markdroth — APPROVED** [review]
Looks great!

**Comment by yousukseung** [comment]
> Looks great!

I removed arg name comments for the matcher instead of adding NOLINT()'s, it looks clang still cannot correctly handle MATCHER macros.

**Pull request merged**
